libvirt/tunnel/tunnel.sh

```diff
@@ -1,98 +1,128 @@
+PORTS="-R $(yq eval '.libvirt.bastion-port' ${filename}):127.0.0.1:$(yq eval '.libvirt.target-port' ${filename})"
+for i in $(seq 0 $(( $CLUSTER_CAPACITY-1 )) ); do
+		PORTS+=" -R $(yq eval '.libvirt-'$ARCH-$CLUSTER_ID-$i'.api.bastion-port' ${filename}):127.0.0.1:$(yq eval '.libvirt-'$ARCH-$CLUSTER_ID-$i'.api.target-port' ${filename}) 
+				 -R $(yq eval '.libvirt-'$ARCH-$CLUSTER_ID-$i'.http.bastion-port' ${filename}):127.0.0.1:$(yq eval '.libvirt-'$ARCH-$CLUSTER_ID-$i'.http.target-port' ${filename}) 
+				 -R $(yq eval '.libvirt-'$ARCH-$CLUSTER_ID-$i'.https.bastion-port' ${filename}):127.0.0.1:$(yq eval '.libvirt-'$ARCH-$CLUSTER_ID-$i'.https.target-port' ${filename}) "
+done
+# echo ${PORTS}
+
+function OC() {	
+	./oc --server https://api.ci.openshift.org --token "${TOKEN}" --namespace "${ENVIRONMENT}" "${@}"
+}
+
+function timestamp() {
+	# With UTC format.  2020-11-04 06:19:24
+	date -u  +"%Y-%m-%d %H:%M:%S"
 }
 
 function port-forward() {
 	while true; do
-		echo "[INFO] Setting up port-forwarding..."
+		echo "$(timestamp) [INFO] Setting up port-forwarding to connect to the bastion..."
 		pod="$( OC get pods --selector component=sshd -o jsonpath={.items[0].metadata.name} )"
-		if ! OC port-forward "${pod}" 2222; then
-			echo "[WARNING] Port-forwarding failed, retrying..."
+		if ! OC port-forward "${pod}" "${1:?Port was not specified}"; then
+			echo "$(timestamp) [WARNING] Port-forwarding failed, retrying..."
 			sleep 0.1
 		fi
 	done
 }
 
+# This opens an ssh tunnel. It uses port 2222 for the ssh traffic.
+# It basically says send traffic from bastion service port to
+# local VM port using port 2222 to establish the ssh connection.
 function ssh-tunnel() {
 	while true; do
-		echo "[INFO] Setting up SSH tunnelling..."
-		if ! ssh -N -T root@127.0.0.1 -p 2222 \
-                        -R ${BASTION_PORTS["LIBVIRT"]}:127.0.0.1:16509 \
-                        -R ${CLUSTER_PORTS["0,API"]}:127.0.0.1:6443   -R ${CLUSTER_PORTS["0,HTTP"]}:127.0.0.1:80    -R ${CLUSTER_PORTS["0,HTTPS"]}:127.0.0.1:443 \
-                        -R ${CLUSTER_PORTS["1,API"]}:127.0.0.1:16443  -R ${CLUSTER_PORTS["1,HTTP"]}:127.0.0.1:10080 -R ${CLUSTER_PORTS["1,HTTPS"]}:127.0.0.1:10443 \
-                        -R ${CLUSTER_PORTS["2,API"]}:127.0.0.1:26443  -R ${CLUSTER_PORTS["2,HTTP"]}:127.0.0.1:20080 -R ${CLUSTER_PORTS["2,HTTPS"]}:127.0.0.1:20443 \
-                        -R ${CLUSTER_PORTS["3,API"]}:127.0.0.1:36443  -R ${CLUSTER_PORTS["3,HTTP"]}:127.0.0.1:30080 -R ${CLUSTER_PORTS["3,HTTPS"]}:127.0.0.1:30443 \
-                        -R ${CLUSTER_PORTS["4,API"]}:127.0.0.1:46443  -R ${CLUSTER_PORTS["4,HTTP"]}:127.0.0.1:40080 -R ${CLUSTER_PORTS["4,HTTPS"]}:127.0.0.1:40443 \
-		; then
-			echo "[WARNING] SSH tunnelling failed, retrying..."
+		echo "$(timestamp) [INFO] Setting up a reverse SSH tunnel to connect bastion port "${1:?Bastion service port and local service port was not specified}"..."
+		if ! ssh -N -T root@127.0.0.1 -p 2222 $1; then
```

Comment:
$1 will only evaluate to -R - you probably want to change that to $@ (and for the echo above)

Reply:
I'm not 100% sure the desired outcome, but if I run this with one of the profiles (profile_ocponz2.yaml for e.g.), the ssh command looks like this:
ssh -N -T root@127.0.0.1 -p 2222 -R 16510:127.0.0.1:16509 -R 6445:127.0.0.1:6445 -R 8081:127.0.0.1:8080 -R 8444:127.0.0.1:8443 -R 6445:127.0.0.1:16445 -R 8081:127.0.0.1:18080 -R 8444:127.0.0.1:18443 -R 6445:127.0.0.1:26445 -R 8081:127.0.0.1:28080 -R 8444:127.0.0.1:28443 -R 6445:127.0.0.1:36445 -R 8081:127.0.0.1:38080 -R 8444:127.0.0.1:38443 -R 6445:127.0.0.1:46445 -R 8081:127.0.0.1:48080 -R 8444:127.0.0.1:48443

Which only works for the first 4 since the ports are duplicated after that, this will link up with the bastion-port in the config, so that needs to not be duplicated for the same host otherwise it can't listen on the same ports. I'm also not sure you want to forward port 6445 to 6445 (the first -R since that won't achieve the desired outcome either)

It would need to look something more like I think, but perhaps Jeremy can weigh in since he's more familiar with the original config.:
ssh -N -T root@127.0.0.1 -p 2222 -R 16510:127.0.0.1:16509 -R 6446:127.0.0.1:6445 -R 8081:127.0.0.1:8080 -R 8444:127.0.0.1:8443 -R 16446:127.0.0.1:16445 -R 18081:127.0.0.1:18080 -R 18444:127.0.0.1:18443 -R 26446:127.0.0.1:26445 -R 28081:127.0.0.1:28080 -R 28444:127.0.0.1:28443 -R 36446:127.0.0.1:36445 -R 38081:127.0.0.1:38080 -R 38444:127.0.0.1:38443 -R 46446:127.0.0.1:46445 -R 48081:127.0.0.1:48080 -R 48444:127.0.0.1:48443